src/mpid/ch4/shm/ipc/src/ipc_mem.h

```diff
@@ -11,33 +11,55 @@
 #include "../xpmem/xpmem_post.h"
 #endif
 
+/*
+=== BEGIN_MPI_T_CVAR_INFO_BLOCK ===
+
+cvars:
+    - name        : MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE
+      category    : CH4
+      type        : int
+      default     : 4096
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        If a send message size is greater than or equal to MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE (in
+        bytes), then enable XPMEM-based single copy protocol for intranode communication. The
+        environment variable is valid only when then XPMEM shmmod is enabled.
+
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
 /* Get local memory handle. No-op if the IPC type is NONE. */
-MPL_STATIC_INLINE_PREFIX int MPIDI_IPC_get_mem_handle(MPIDI_SHM_IPC_type_t ipc_type,
-                                                      void *vaddr, size_t size,
-                                                      MPIDI_IPC_mem_handle_t * mem_handle)
+MPL_STATIC_INLINE_PREFIX int MPIDI_IPC_get_mem_attr(const void *vaddr, size_t size,
+                                                    int rank, MPIR_Comm * comm,
+                                                    MPIDI_IPC_mem_attr_t * attr)
 {
     int mpi_errno = MPI_SUCCESS;
 
-    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_IPC_GET_MEM_HANDLE);
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_IPC_GET_MEM_HANDLE);
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_IPC_GET_MEM_ATTR);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_IPC_GET_MEM_ATTR);
 
-    switch (ipc_type) {
 #ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
-        case MPIDI_SHM_IPC_TYPE__XPMEM:
-            mpi_errno = MPIDI_IPC_xpmem_get_mem_handle(vaddr, size, &mem_handle->xpmem);
-            break;
+    if (MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE > -1 &&
+        size >= MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE && rank != comm->rank) {
+        mpi_errno = MPIDI_IPC_xpmem_get_mem_handle(vaddr, size, &attr->mem_handle.xpmem);
+        MPIR_ERR_CHECK(mpi_errno);
+        attr->ipc_type = MPIDI_SHM_IPC_TYPE__XPMEM;
+    } else
 #endif
```

Comment:
If we provide empty version of XPMEM mem_handle routines, we can make the code in `ipc_mem.h` cleaner. How about define:
```
#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
#define MPIDI_XPMEM_ENABLE_LMT (MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE > -1 )
#else
#define MPIDI_XPMEM_ENABLE_LMT (0)
#endif 
```
Then we can avoid macro-level check in ipc_mem.h

Reply:
I think this is fine. I will make the change.

Reply:
@minsii I have added another commit to make this last batch of changes. I will squash this with the previous if you are happy with it.